tools/cldr-apps/src/main/java/org/unicode/cldr/web/SearchManager.java

```diff
@@ -0,0 +1,237 @@
+            // TODO: could potentially cache the results here.
+        }
+
+        @Override
+        public String toString() {
+            return String.format("[SearchResponse#%s: %s, %s, #=%d]",
+                token,
+                isComplete ? "complete" : "incomplete",
+                isOngoing ? "ongoing" : "stopped",
+                results.size());
+        }
+    }
+
+    /**
+     * A search in progress
+     */
+    private class Search implements Callable<Search> {
+
+        private final SearchRequest request;
+        private final SearchResponse response = new SearchResponse();
+        private final String locale;
+        private Future<Search> future;
+
+        Search(SearchRequest request, String locale) {
+            this.locale = locale;
+            this.request = request;
+            this.response.lastUpdated = this.response.searchStart = new Date();
+            this.response.isOngoing = true;
+            this.response.isComplete = false;
+            this.response.token = CookieSession.newId();
+            logger.fine(()-> String.format("%s: l=%s, q='%s'",
+                this.response,
+                this.locale,
+                this.request.value));
+        }
+
+        public void begin() {
+            // TODO: could look up cached results here.
+
+            this.future = SurveyThreadManager
+                .getExecutorService()
+                .submit(this);
+        }
+
+        @Override
+        public Search call() throws Exception {
+            // first, simple search of exact values
+            CLDRFile file = factory.make(locale, false);
+            Set<String> xresult = new TreeSet<>();
+            file.getPathsWithValue(request.value, "", null, xresult);
+            for (final String xpath : xresult) {
+                // Skip if this isn’t found "here"
+                if(!file.isHere(xpath)) continue;
+                // Add incrementally. A user may get a partial result if they request before we are done.
+                response.addResult(new SearchResult(
+                    xpath, request.value, locale
+                ));
+            }
+
+            // All done (for now!)
+            response.complete();
+
+            return this;
+        }
+
+        public void stop() {
+            future.cancel(true);
+            response.complete();
+        }
+    }
+
+    private Factory factory;
+
+    final Cache<Object, Object> searches = CacheBuilder.newBuilder()
+        .expireAfterWrite(10, TimeUnit.MINUTES)
+        .build();
+
+    private SearchManager(Factory f) {
+        this.factory = f;
+    }
+
+    /**
+     * Factory for the SearchManager
+     * @param f
+     * @return
+     */
+    public static SearchManager forFactory(Factory f) {
+        return new SearchManager(f);
+    }
+
+    /**
+     * Start up a new search
+     * @param request
+     * @return
+     */
+    public SearchResponse newSearch(final SearchRequest request, final String locale) {
```

Comment:
should this be synchronized on searches or would that be superfluous?

Reply:
superfluous.  the Cache is safe and only one thread has knowledge of that Search.